File: solutions/haskell/list-ops/1/src/ListOps.hs
module ListOps
  ( length,
    reverse,
    map,
    filter,
    foldr,
    foldl',
    (++),
    concat,
  )
where

import Prelude hiding
  ( concat,
    filter,
    foldr,
    length,
    map,
    reverse,
    (++),
  )

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' f z xs = if null xs then z else let z' = f z (head xs) in seq z' (foldl' f z' (tail xs))

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr f z xs = if null xs then z else f (head xs) (foldr f z (tail xs))

length :: [a] -> Int
length xs = if null xs then 0 else 1 + length (tail xs)

reverse :: [a] -> [a]
reverse xs = if null xs then [] else (last xs) : reverse (init xs)

map :: (a -> b) -> [a] -> [b]
map f xs = if null xs then [] else f (head xs) : map f (tail xs)

filter :: (a -> Bool) -> [a] -> [a]
filter p xs =
  if null xs
    then []
    else
      if p $ head xs then (head xs) : filter p (tail xs) else filter p $ tail xs

(++) :: [a] -> [a] -> [a]
xs ++ ys = if null xs then ys else head xs : (tail xs ++ ys)

concat :: [[a]] -> [a]
concat xss = if null xss then [] else head xss ++ concat (tail xss)

File: solutions/fsharp/list-ops/1/ListOps.fs
module ListOps

let rec rev_aux l r =
    match l with
    | [] -> r
    | x :: xs -> rev_aux xs (x::r)
let reverse list = rev_aux list []
let rec foldl folder state list = 
    match list with
    | [] -> state
    | x :: xs -> foldl folder (folder state x) xs

let rec foldr folder state list = foldl folder state (reverse list)

let rec length_aux l n = 
    match l with
    | [] -> n
    | _ :: xs -> length_aux xs (n + 1)
let length list = length_aux list 0

let rec map_aux f l r = 
    match l with
    | [] -> reverse r
    | x :: xs -> map_aux f xs ((f x) :: r)
let map f list = map_aux f list []

let rec fil_aux f l r = 
    match l with
    | [] -> reverse r
    | x :: xs ->
        if f x then
            fil_aux f xs (x::r)
        else
            fil_aux f xs r
let filter f list = fil_aux f list []

let rec ap_aux xs ys = 
    match ys with 
    | [] -> reverse xs
    | y :: ys -> ap_aux (y :: xs) ys
let append xs ys = ap_aux (reverse xs) ys

let concat xs = foldl (fun acc a -> append acc a) [] xs